lunamark/writer/generic.lua

```diff
@@ -255,8 +255,12 @@ function M.new(options)
   end
 
   -- (Block) Div with attributes `attr`
-  function W.div(s)
-    return s
+  function W.div_begin(_)
```

Comment:
TBH, I don't like this `div_begin`/`div_end` split to the writer API, because it loses the hierarchy, flattening it to a linear
(`div_begin()`,  content, ... content, `div_end()`) instead of the structured (`div`(content, ... content)). I see several drawbacks:

 - It makes the task of the writer(s) harder, when they want/need to reconstruct the hierarchy.
 - It introduces a difference with how Spans are handled (they keep the hierarchy)
 - It deviates from Pandoc, which keeps the hierachy too, e.g. in terms of [Pandoc AST](https://hackage.haskell.org/package/pandoc-types-1.22.2.1/docs/Text-Pandoc-Definition.html#t:Block) a "Div Attr [Blocks]" constructs, e.g.
   ```
   $ pandoc -t json
   ::: { .class }
   Content
   :::
   ```
   Gives (spaces and carriage returns added for readability):
   ```
   {
     "pandoc-api-version": [1, 22, 2],
     "meta": {},
     "blocks": [ {
         "t": "Div",
         "c": [ [ "", ["class"], [] ],   [ {
            "t": "Para",
            "c": [ { "t": "Str",   "c": "Content" } ]
         } ]
      ] } ]
   }
   ```

Reply:
N.B. Hierarchy vs. Flat representation is also some a good question if we ever want one day to support the `native_divs` with HTML...
```
$ pandoc -t json # native_divs is enabled by default in Pandoc
<div>
  Text
</div>
``
Gives `{ ... [{"t":"Div","c":[...],[{"t":"Para","c":[{"t":"Str","c":"Text"}]}]]}]}
(hierarchical)

Whereas 
```
$ pandoc -t json -f markdown-native_divs
<div>
  Text
</div>
```
Gives `{ ... [{"t":"RawBlock","c":["html","<div>"]},{"t":"Plain","c":[{"t":"Str","c":"Text"}]},{"t":"RawBlock","c":["html","</div>"]}]}`
(flat)

Reply:
> It makes the task of the writer(s) harder, when they want/need to reconstruct the hierarchy

You could argue that headings introduce a hierarchy of sections, which the writer also has to keep track of. As a matter of fact, that's what the TeX writer of witiko/markdown does.

> It introduces a difference with how Spans are handled (they keep the hierarchy)

Spans are inline elements, which makes all the difference in terms of parsing. If you have an idea how to buffer the entire div at once before passing it to the div writer, I will appreciate the advice because I will admit that I don't.

> Hierarchy vs. Flat representation is also some a good question if we ever want one day to support the native_divs with HTML ...

Interestingly enough, Lunamark already correctly parses nested block-level HTML elements. The code is a bit of a puzzle, but it correctly does the recursive matching of tags:

https://github.com/jgm/lunamark/blob/d93b9e3de3bfce0fd4727b65e6d8d2d9259c06cb/lunamark/reader/markdown.lua#L469-L491

The magic is in the `V(1)` pattern, which corresponds to the `parsers.openelt_exact(s) ... parsers.closeelt_exact(s)` pattern itself, see the [LPEG docs][1]. Remove it and Lunamark will fail to recognize nested structures such as `<div>foo<div>bar</div>baz</div>`. Keep it and it works as a charm.

Perhaps we could use a similar technique to parse both fenced divs and native divs at once. However, this seems much more difficult, because whereas block-level HTML elements contain raw text (similarly fenced code), we have to use the full markdown grammar when parsing ahead fenced divs and native divs, so that we don't get fooled by `:::` in a fenced code and similar tricks.

 [1]: http://www.inf.puc-rio.br/~roberto/lpeg/#grammar

Reply:
Unless you object, I would merge this as a bugfix, so that we recognize fenced divs correctly, and we can work on improving the writer side of things later. Seems less urgent.

Reply:
> This will allow us to fenced divs the same way Pandoc does.
No, Pandoc since keeps the hierarchy and we are not, here.

I really don't like it. I mean, the rationale for keeping the structure hierarchy ought to be sound. Would you be happy if, say `writer.singlequoted(content)` was actually `writer.singlequote_begin()` and `writer.singlequote_end()`, each responsible for outputting the right quote, and content was split apart? Or fenced code blocks, too.

The Markdown file has a inherent structure. My feeling is that if our writer API is not able to get it right, it's probably bad. (Or at least, say: worst than _not_ following Pandoc's mess *grins* and assume same numbers of separators for opening and closing divs ^^)

Reply:
> You could argue that for example headings introduce a hierarchy of sections, which the writer may also need to keep track of

I would argue so, theoretically. But Pandoc does lose the possible hierarchy here too, so at least (or worst) we are on par with it.

Reply:
> But Pandoc does lose the possible hierarchy here too, so at least (or worst) we are on par with it.

That's one of my points. Another one is that although `div_begin()` and `div_end()` are ugly, they are still much nicer than `heading()` for a writer that keeps track of sections.

> No, Pandoc since keeps the hierarchy and we are not, here.

The output of the writer is consistent with Pandoc even though we forget the hierarchy internally.

> I really don't like it.

I don't like it either, but it works fine, because none of our writers really need the hierarchical structure. I would like to know how to parse nested divs without losing the hierarchical structure, but I worry that even if we do get it right, the code footprint will be much larger without many tangible benefits.

Reply:
I have a version using the `V(1)` trick (I was looking a that too, hence my mentioning the HTML stuff) and it does passes your test. I need to clean it up, though.